macrosynergy/panel/adjust_weights.py

```diff
@@ -0,0 +1,306 @@
+        (method, "method", Callable),
+        (params, "param", dict),
+        (cids, "cids", (list, type(None))),
+        (start, "start", (str, type(None))),
+        (end, "end", (str, type(None))),
+    ]:
+        if not isinstance(_var, _type):
+            raise TypeError(f"{_name} must be a {_type}, not {type(_var)}")
+
+    if cids is not None and (
+        not all(isinstance(cid, str) for cid in cids) or len(cids) == 0
+    ):
+        raise TypeError("`cids` must be a None(default) or a non-empty list of strings")
+
+
+def adjust_weights_backend(
+    df_weights_wide: pd.DataFrame,
+    df_adj_zns_wide: pd.DataFrame,
+    method: Callable,
+    params: Dict[str, Any] = {},
+) -> pd.DataFrame:
+    """
+    Backend function for adjust_weights. Applies the `method` function to the weights and
+    multiplies the result by the adjustment factors, and by the parameter `param`.
+    Expects the input DataFrames to be in wide format, with the same columns AND index
+    (see macrosynergy.panel.adjust_weights.split_weights_adj_zns).
+
+    Parameters
+    ----------
+    df_weights_wide : pd.DataFrame
+        DataFrame with weights in wide format.
+
+    df_adj_zns_wide : pd.DataFrame
+        DataFrame with adjustment factors in wide format.
+
+    method : Callable
+        Function that will be applied to the weights to adjust them.
+
+    params : Dict[str, Any], optional
+        Parameters to be passed to the method function. Default is {}.
+
+    Returns
+    -------
+    pd.DataFrame
+        DataFrame with the adjusted weights.
+    """
+
+    assert set(df_weights_wide.columns) == set(df_adj_zns_wide.columns)
+    assert set(df_weights_wide.index) == set(df_adj_zns_wide.index)
+
+    dfw_result = df_weights_wide * df_adj_zns_wide.map(method, **params)
+
+    return dfw_result
+
+
+def split_weights_adj_zns(
+    df: QuantamentalDataFrame, weights: str, adj_zns: str
+) -> Tuple[pd.DataFrame, pd.DataFrame]:
+    """
+    Splits the input DataFrame into two DataFrames, one containing the weights and the
+    other containing the adjustment factors.
+
+    Parameters
+    ----------
+    df : QuantamentalDataFrame
+        DataFrame containing the weights and adjustment factors.
+
+    weights : str
+        Name of the xcat to be used as weights.
+
+    adj_zns : str
+        Name of the z-n score xcat to be used as adjustment factors.
+
+    Returns
+    -------
+    Tuple[pd.DataFrame, pd.DataFrame]
+        Tuple containing two wide DataFrames (one for weights and one for adjustment
+        factors), with one column per cid.
+    """
+
+    df_weights = df.loc[df["xcat"] == weights]
+    df_adj_zns = df.loc[df["xcat"] == adj_zns]
+
+    df_weights_wide = QuantamentalDataFrame(df_weights).to_wide()
+    df_adj_zns_wide = QuantamentalDataFrame(df_adj_zns).to_wide()
+
+    # convert the weights to absolute values (no negative weights in this use case)
+    df_weights_wide = df_weights_wide.abs()
```

Comment:
@Magnus167 to me this should be a ValueError, rather than force it to be positive. If we have a value of -0.5, we don't know if the weight is actually 0.5 or if it is a rounding error.

@sandresen1 was this your point about absolute as well?

In general, we should have a check for:

1. df_weights all N/A or positive (between either 0 to 100 or 0 to 1).
2. The transformed weights being similar.